### PR TITLE
Feature/mcd v2automation depricator

### PR DESCRIPTION
**Pull request by kjaredic**

### CODE REVIEW

File: contracts/exchangeV3/wrappersV3/MockWrapper.sol [hunk 1]
```diff
@@ -0,0 +1,57 @@
...
+import "../../interfaces/curve/ISwaps.sol";
+import "../../interfaces/IERC20.sol";
+import "../../DS/DSMath.sol";
+import "../../auth/AdminAuth.sol";
+import "../../utils/SafeERC20.sol";
+import "../../utils/TokenUtils.sol";
+import "./helpers/WrapperHelper.sol";
+
+/// @title DFS exchange wrapper used for mocking in tests
+contract MockWrapper is DSMath, IExchangeV3, AdminAuth, WrapperHelper {
```
Comment (yippee-ki-yay) [inline contracts/exchangeV3/wrappersV3/MockWrapper.sol:14]: Good idea to mock this, in a new push to main branch we now have `contracts/mocks` folder where we can put mock contracts to separate them from production contracts, shift there and rename the contract to `MockExchangeWrapper` to be more descriptive.

File: test/strategies/mcd/mcd-tests.js [hunk 2]
```diff
@@ -133,45 +148,31 @@ const mcdBoostStrategyTest = async () => {
...
+            await redeploy('McdFLBoostComposite');
             strategyTriggerView = await redeploy('StrategyTriggerView');
             await addBotCaller(botAcc.address);
 
             proxy = await getProxy(senderAcc.address);
         });
 
-        it('... should create 2 boost strategies and create a bundle', async () => {
+        it('... should create 4 boost strategies and create a bundle', async () => {
             const boostStrategy = createMcdBoostStrategy();
```
Comment (yippee-ki-yay) [inline test/strategies/mcd/mcd-tests.js:160]: Switch normal recipe strategies, to flBalancer as well for better gas comparison

File: contracts/actions/mcd/McdFLRepayComposite.sol [hunk 3]
```diff
@@ -0,0 +1,195 @@
...
+            uint256 frobAmount = convertTo18(_repayParams.joinAddr, repayAmount);
+
+            IManager(MCD_MANAGER_ADDR).frob(
+                _repayParams.vaultId,
+                -toPositiveInt(frobAmount),
+                paybackAmountNormalized
+            );
+            IManager(MCD_MANAGER_ADDR).flux(_repayParams.vaultId, address(this), frobAmount);
+            // withdraw the tokens from Join
+            IJoin(_repayParams.joinAddr).exit(address(this), repayAmount);
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdFLRepayComposite.sol:166]: If FL has fee this will return amount without fee and will not be able to repay FL

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 4]
```diff
@@ -0,0 +1,172 @@
...
+        return uint8(ActionType.CUSTOM_ACTION);
+    }
+
+    function _parseCompositeParams(bytes memory _calldata) internal pure returns (RepayParams memory params) {
+        params = abi.decode(_calldata, (RepayParams));
+    }
+
+    /// @dev workaround for dfsSell expecting context to be proxy
+    function owner() external pure returns (address) {
+        return address(0);
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdRepayComposite.sol:170]: Can this field be 0x0 and 0x and other exchanges still working normally/tracking not changed?

Comment (yippee-ki-yay) [reply contracts/actions/mcd/McdRepayComposite.sol:170]: And isn't this call in the context of dsproxy already?

File: contracts/actions/mcd/McdFLBoostComposite.sol [hunk 5]
```diff
@@ -0,0 +1,190 @@
...
+        return uint8(ActionType.CUSTOM_ACTION);
+    }
+
+    function _parseCompositeParams(bytes memory _calldata) internal pure returns (BoostParams memory params) {
+        params = abi.decode(_calldata, (BoostParams));
+    }
+
+    /// @dev workaround for dfsSell expecting context to be proxy
+    function owner() external pure returns (address) {
+        return address(0);
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdFLBoostComposite.sol:188]: Might cause exchange issue if 0x0

File: contracts/actions/mcd/McdRepayComposite.sol (hunk 4 above)
Comment (kjaredic) [reply contracts/actions/mcd/McdRepayComposite.sol]: Negligent use of copy paste, fixed.

File: contracts/actions/mcd/McdSubProxy.sol [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+        repaySub.triggerData[0] = triggerData;
+
+        repaySub.subData =  new bytes32[](2);
+        repaySub.subData[0] = bytes32(_user.vaultId);
+        repaySub.subData[1] = bytes32(uint256(_user.targetRatioRepay));
+        
+    }
+
+    /// @notice Formats a StrategySub struct to a Boost bundle from the input data of the specialized aave sub
+    function formatBoostSub(McdSubData memory _user) public pure returns (StrategySub memory repaySub) {
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdSubProxy.sol:142]: StrategySub memory repaySub -> boostSub and comments are saying aave instead of mcd

File: contracts/actions/mcd/McdSubProxy.sol [hunk 7]
```diff
@@ -0,0 +1,155 @@
...
+
+pragma solidity =0.8.10;
+
+import "../../auth/AdminAuth.sol";
+import "../../auth/ProxyPermission.sol";
+import "../../core/strategy/SubStorage.sol";
+
+/// @title Subscribes users to boost/repay strategies in an L2 gas efficient way
+contract McdSubProxy is StrategyModel, AdminAuth, ProxyPermission, CoreHelper {
+    uint64 public constant REPAY_BUNDLE_ID = 3;
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdSubProxy.sol:11]: Outdated

File: test/strategies.js [hunk 8]
```diff
@@ -1336,6 +1388,56 @@ const createFlMcdBoostStrategy = () => {
...
+
+    const mcdRatioTrigger = new dfs.triggers.MakerRatioTrigger('0', '0', '0');
+    mcdBoostStrategy.addTrigger(mcdRatioTrigger);
+
+    const boostCompositeAction = new dfs.actions.maker.MakerFLBoostCompositeAction(
+        '&vaultId',
+        '%joinAddr',
+        '%gasUsed',
+        formatExchangeObj(
+            '%daiAddr',
```
Comment (yippee-ki-yay) [inline test/strategies.js:1429]: We can hardcode dai address as well as we did in other strategies

File: contracts/actions/mcd/McdFLRepayComposite.sol [hunk 9]
```diff
@@ -0,0 +1,208 @@
...
+        uint256 paybackAmount = _takeFee(GasFeeTakerParams(
+            _repayParams.gasUsed,
+            DAI_ADDR,
+            exchangedAmount,
+            0
+        ));
+
+        // Withdraw collateral and payback debt
+        {
+            // if paybackAmount is higher than current debt, repay all debt and send remaining dai to proxy
```
Comment (yippee-ki-yay) [inline contracts/actions/mcd/McdFLRepayComposite.sol:153]: comment wrong, and send remaining dai to proxy owner

**Review by yippee-ki-yay — CHANGES_REQUESTED** [review]
Overall well tested and optimized, needs a few minor changes for approval written in comments

File: contracts/actions/fee/GasFeeTaker.sol [hunk 10]
```diff
@@ -26,53 +26,56 @@ contract GasFeeTaker is ActionBase, GasFeeHelper {
...
+    function executeActionDirect(bytes memory _callData) public payable virtual override {}
 
     /// @inheritdoc ActionBase
     function actionType() public pure virtual override returns (uint8) {
         return uint8(ActionType.FEE_ACTION);
     }
 
-    function parseInputs(bytes memory _callData) public pure returns (Params memory inputData) {
-        inputData = abi.decode(_callData, (Params));
+    function parseInputsGasFeeTaker(bytes memory _callData) public pure returns (GasFeeTakerParams memory inputData) {
```
Comment (majkic99) [inline contracts/actions/fee/GasFeeTaker.sol:78]: any specific reason for changing function and struct names
Full name of the struct would go from GasFeeTaker.Params to GasFeeTaker.GasFeeTakerParams and parseInputs is called that way in every action, should we change it in others, or revert this change here for consistency?

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 11]
```diff
@@ -0,0 +1,213 @@
...
+        (bytes memory logData, ) = _boost(boostParams);
+        logger.logActionDirectEvent("McdBoostComposite", logData);
+    }
+
+    /// @notice Executes boost logic
+    function _boost(BoostParams memory _boostParams)
+        internal
+        returns (bytes memory logData, uint256 supplyAmount)
+    {
+        assert(_boostParams.exchangeData.srcAddr == DAI_ADDR);
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:101]: should be require or a custom error, not assert

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 12]
```diff
@@ -0,0 +1,213 @@
...
+        }
+
+        // if we're using flashloan draw debt and repay fl
+        if (_boostParams.flAddr != address(0)) {
+            uint256 daiAmount = _drawDebt(_boostParams, urn, ilk);
+            DAI_ADDR.withdrawTokens(_boostParams.flAddr, daiAmount);
+        }
+
+        // check if collateral is higher, only if part of strategy
+        if (_boostParams.gasUsed != 0) {
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:160]: in which case is it possible for collateral to be less than what we begin with? From my understanding we're supplying once and not withdrawing at any point?

Comment (majkic99) [reply contracts/actions/mcd/McdBoostComposite.sol:160]: maybe check if ratio is lower than before always?

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 13]
```diff
@@ -0,0 +1,213 @@
...
+                    exchangedAmount,
+                    MAX_DFS_FEE
+                )
+            );
+        } else {
+            supplyAmount = exchangedAmount;
+        }
+
+        // Supply collateral
+        {
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:143]: maybe create a helper function like _drawDebt (_supplyColl) to keep code cleaner

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 14]
```diff
@@ -0,0 +1,213 @@
...
+import "../ActionBase.sol";
+import "../exchange/DFSSell.sol";
+import "../fee/GasFeeTaker.sol";
+
+import "../../interfaces/mcd/IManager.sol";
+import "../../interfaces/mcd/IDaiJoin.sol";
+import "../../interfaces/mcd/IJug.sol";
+
+import "../../utils/TokenUtils.sol";
+import "../../core/strategy/StrategyModel.sol";
```
Comment (majkic99) [inline contracts/actions/mcd/McdBoostComposite.sol:13]: unused import

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 15]
```diff
@@ -0,0 +1,200 @@
+// SPDX-License-Identifier: MIT
+pragma solidity 0.8.10;
+
+import "../ActionBase.sol";
+import "../exchange/DFSSell.sol";
+import "../fee/GasFeeTaker.sol";
+
+import "../../interfaces/balancer/IFlashLoanRecipient.sol";
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:8]: lines 8,9,15,18 unused imports

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 16]
```diff
@@ -0,0 +1,200 @@
...
+        override(ActionBase, DFSSell, GasFeeTaker)
+    {
+        RepayParams memory repayParams = _parseCompositeParams(_callData);
+        (bytes memory logData, ) = _repay(repayParams);
+        logger.logActionDirectEvent("McdRepayComposite", logData);
+    }
+
+    /// @notice Executes repay logic
+    function _repay(RepayParams memory _repayParams) internal returns (bytes memory logData, uint256 paybackAmount) {
+        assert(_repayParams.exchangeData.destAddr == DAI_ADDR);
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:102]: replace with require or custom error

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 17]
```diff
@@ -0,0 +1,200 @@
...
+        if (_repayParams.gasUsed != 0) {
+            paybackAmount = _takeFee(
+                GasFeeTakerParams(_repayParams.gasUsed, DAI_ADDR, exchangedAmount, MAX_DFS_FEE)
+            );
+        } else {
+            paybackAmount = exchangedAmount;
+        }
+
+        // Payback debt
+        {
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:135]: maybe put this in helper function

File: contracts/actions/mcd/McdRepayComposite.sol [hunk 18]
```diff
@@ -0,0 +1,200 @@
...
+                urn,
+                ilk
+            );
+
+            IManager(MCD_MANAGER_ADDR).frob(_repayParams.vaultId, 0, paybackAmountNormalized);
+        }
+
+        if (_repayParams.flAddress != address(0)) {
+            _withdrawColl(_repayParams, repayAmount);
+            _repayParams.exchangeData.srcAddr.withdrawTokens(_repayParams.flAddress, repayAmount);
```
Comment (majkic99) [inline contracts/actions/mcd/McdRepayComposite.sol:159]: this would only work with flashloan providers that don't have any fee, how do we handle if we use a provider with fee? Boost is maybe ok because we flashloan DAI and there should always be free DAI fl providers, but we can't say that for certain for all of coll types

File: contracts/actions/mcd/McdSubProxy.sol [hunk 19]
```diff
@@ -0,0 +1,162 @@
+// SPDX-License-Identifier: MIT
+
+pragma solidity =0.8.10;
+
+import "../../auth/AdminAuth.sol";
+import "../../auth/ProxyPermission.sol";
+import "../../core/strategy/SubStorage.sol";
+import "../../utils/helpers/UtilHelper.sol";
+
+/// @title Subscribes users to boost/repay strategies in an L2 gas efficient way
```
Comment (majkic99) [inline contracts/actions/mcd/McdSubProxy.sol:10]: l2 gas efficient + mcd?

File: contracts/exchangeV3/DFSExchangeCore.sol [hunk 20]
```diff
@@ -204,7 +201,7 @@ contract DFSExchangeCore is DFSExchangeHelper, DSMath, DFSExchangeData, Exchange
                 feeAmount = _amount / 10;
             }
 
-            address walletAddr = feeRecipient.getFeeAddr();
+            address walletAddr = FeeRecipient(FEE_RECIPIENT_ADDRESS).getFeeAddr();
```
Comment (majkic99) [inline contracts/exchangeV3/DFSExchangeCore.sol:204]: unnecessary change that would mean we should redeploy dfs sell?

File: contracts/actions/mcd/McdRepayComposite.sol (hunk 17 above)
Comment (yippee-ki-yay) [reply contracts/actions/mcd/McdRepayComposite.sol:135]: less gas efficient, that's why it's all block scoped

File: contracts/actions/mcd/McdBoostComposite.sol (hunk 12 above)
Comment (yippee-ki-yay) [reply contracts/actions/mcd/McdBoostComposite.sol:160]: This is mainly done as sanity checks, if something goes wrong with the execution, if i check if the ratio is lower that might not mean that the boost went through correctly, might have just withdrawn funds, sold and got stuck on proxy. It seemed better to check if collateral has increased.

File: contracts/exchangeV3/DFSExchangeCore.sol (hunk 20 above)
Comment (yippee-ki-yay) [reply contracts/exchangeV3/DFSExchangeCore.sol:204]: Action inherits 2 contracts with feeRecipient, does not mean we need to redeploy dfs sell as the code is inherited into the action

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 21]
```diff
@@ -0,0 +1,229 @@
...
+        );
+
+        boostParams.flAmount = _parseParamUint(
+            boostParams.flAmount,
+            _paramMapping[2],
+            _subData,
+            _returnValues
+        );
+
+        boostParams.exchangeData.srcAddr = _parseParamAddr(
```
Comment (kjaredic) [inline contracts/actions/mcd/McdBoostComposite.sol:74]: seems redundant to map this as we already have a check below forcing this to be DAI

Comment (kjaredic) [reply contracts/actions/mcd/McdBoostComposite.sol:74]: similar for RepayComposite

File: contracts/actions/mcd/McdBoostComposite.sol [hunk 22]
```diff
@@ -0,0 +1,229 @@
...
+            _returnValues
+        );
+
+        boostParams.exchangeData.srcAmount = _parseParamUint(
+            boostParams.exchangeData.srcAmount,
+            _paramMapping[5],
+            _subData,
+            _returnValues
+        );
+
```
Comment (kjaredic) [inline contracts/actions/mcd/McdBoostComposite.sol:93]: shouldn't TargetRatio also be mapped as it is defined by each subscription?

**Review by kjaredic — COMMENTED** [review]
looks good, small issues relating to param mapping need to be addressed

**Review by kjaredic — COMMENTED** [review]
should remove composite action tests from test/mcd as they are no longer valid

File: contracts/actions/mcd/McdBoostComposite.sol (hunk 22 above)
Comment (yippee-ki-yay) [reply contracts/actions/mcd/McdBoostComposite.sol:93]: Must keep variable to handle nextPrice injection which will not hit the current targetRatio
